ext/openssl/openssl.c

```diff
@@ -1018,9 +1099,47 @@ PHP_MINIT_FUNCTION(openssl)
 {
 	char * config_filename;
 
-	le_key = zend_register_list_destructors_ex(php_openssl_pkey_free, NULL, "OpenSSL key", module_number);
-	le_x509 = zend_register_list_destructors_ex(php_openssl_x509_free, NULL, "OpenSSL X.509", module_number);
-	le_csr = zend_register_list_destructors_ex(php_openssl_csr_free, NULL, "OpenSSL X.509 CSR", module_number);
+	zend_class_entry ce;
+	INIT_CLASS_ENTRY(ce, "X509Certificate", class_X509Certificate_methods);
+	x509_certificate_ce = zend_register_internal_class(&ce);
+	x509_certificate_ce->ce_flags |= ZEND_ACC_FINAL | ZEND_ACC_NO_DYNAMIC_PROPERTIES;
+	x509_certificate_ce->create_object = x509_certificate_create_object;
+	x509_certificate_ce->serialize = zend_class_serialize_deny;
+	x509_certificate_ce->unserialize = zend_class_unserialize_deny;
+
+	memcpy(&x509_certificate_object_handlers, &std_object_handlers, sizeof(zend_object_handlers));
+	x509_certificate_object_handlers.offset = XtOffsetOf(x509_certificate_object, std);
+	x509_certificate_object_handlers.free_obj = x509_certificate_free_obj;
+	x509_certificate_object_handlers.get_constructor = x509_certificate_get_constructor;
+	x509_certificate_object_handlers.clone_obj = NULL;
+
+	zend_class_entry csr_ce;
+	INIT_CLASS_ENTRY(csr_ce, "X509CertificateSigningRequest", class_X509CertificateSigningRequest_methods);
+	x509_request_ce = zend_register_internal_class(&csr_ce);
+	x509_request_ce->ce_flags |= ZEND_ACC_FINAL | ZEND_ACC_NO_DYNAMIC_PROPERTIES;
```

Comment:
this reminds me that instantiation should be also prohibited for all classes (meaning something like private constructor)

Reply:
Yes, that's right! The constructors currently emit an exception with the `Cannot directly construct...` text, which is in line what we did for other resource migrations.